azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
+                                self.name, self.state, self.pipeline_root.connected
+                            )
+                        )
+
+            elif isinstance(event, pipeline_events_base.DisconnectedEvent):
+                with self.state_lock:
+                    # UNEXPECTED DISCONNECTION (i.e. Connection has been lost)
+                    if self.state is ReconnectState.CONNECTED:
+                        # When we get disconnected, we try to reconnect as soon as we can. We set a
+                        # timer here that will start the process in another thread because we don't
+                        # want to hold up the event flow
+                        logger.debug(
+                            "{}({}): State changes CONNECTED -> DISCONNECTED. Attempting to reconnect".format(
+                                self.name, event.name
+                            )
+                        )
+                        self.state = ReconnectState.DISCONNECTED
+                        self._start_reconnect_timer(0.01)
 
+                    # EXPECTED DISCONNECTION (DisconnectOperation was previously issued)
+                    elif self.state is ReconnectState.DISCONNECTING:
+                        # No reconnect timer will be created.
+                        logger.debug(
+                            "{}({}): State changes DISCONNECTING -> DISCONNECTED. Not attempting to reconnect (User-initiated disconnect)".format(
+                                self.name, event.name
+                            )
+                        )
+                        self.state = ReconnectState.DISCONNECTED
+
+                    # EXPECTED DISCONNECTION (Reauthorization process)
+                    elif self.state is ReconnectState.REAUTHORIZING:
+                        # ReconnectState will remain REAUTHORIZING until completion of the process
+                        # upon re-establishing the connection
+
+                        # NOTE: There is a ~small~ chance of a false positive here if an unexpected
+                        # disconnection occurs while a ReauthorizationOperation is in flight.
+                        # However, it will sort itself out - the ensuing connect that ocurrs as part
+                        # of the reauthorization will restore connection (no harm done) of it will
+                        # fail, at which point the failure was a result of a manual operation and
+                        # reconnection is not supposed to occur. So either way, we end up where we want
+                        # to be despite the false positive - just be aware that this can happen.
+                        logger.debug(
+                            "{}({}): Not attempting to reconnect (Reauthorization in progress)".format(
+                                self.name, event.name
+                            )
+                        )
+
+                    # BAD STATE (this block should not be reached)
```

Comment:
># BAD STATE (this block should not be reached) [](http://example.com/codeflow?start=20&length=46)

Again, nothing guaranting that it won't be reached.  What if you're in ReconnectState.CONNECTING and you get a DisconnectedEvent?  That's not very hard to imagine.

Reply:
This is one of the few changes I had yet to push - we now do change the state to DISCONNECTED, but still consider this to be irregular and log a warning as such.

It should not be possible to receive a DisconnectedEvent while in a CONNECTING state, because a DisconnectedEvent should only fire when a connection has been established, and then lost. Once a connection is established in transport, the first thing that happens is a ConnectedEvent fires up, which will immediately put us in a CONNECTED state. It should not be possible for the ReconnectStage to receive an ensuing DisconnectedEvent prior to receiving the ConnectedEvent.

I say "should", but I'm not going to completely rule out the possibility. However if we come across such possibility, we should seek to eliminate such aberrant behavior.

Reply:
Its a tricky problem.  My biggest area of uncertainty is around queueing of events.  The network stack, or Paho, or our pipeline thread could queue events, so the code isn't 100% guaranteed to handle events in the order that they are received.

For example, in pipeline_stages_mqtt.py, `_on_mqtt_connected` is decorated with `invoke_on_pipeline_thread_nowait`, so the call to `_on_mqtt_connected` can get queued up to run in the pipeline thread the next time it's avaialable, but that might be a while if the pipeline thread is busy doing something (like completing a bunch of pending operatings). 

That being said, getting stuck in an intermediate state is the biggest problem because it can block ops from going down.  I think your code favors leaving intermediate states, and that's a good thing.